larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.cc

```diff
@@ -0,0 +1,129 @@
+/**
+ *  @file   larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.cc
+ *
+ *  @brief  Implementation of the discrete probability helper helper class.
+ *
+ *  $Log: $
+ */
+
+#include "larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.h"
+
+namespace lar_content
+{
+
+template <typename T>
+float LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromPermutationTest(const T &t1, const T &t2, 
+    std::mt19937 &randomNumberGenerator, const size_t nPermutations)
+{
+    if (1 > nPermutations)
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    const float rNominal(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(t1,t2));
+
+    int nExtreme(0);
+    for (unsigned int iPermutation = 0; iPermutation < nPermutations; ++iPermutation)
+    {
+        const float rRandomised(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(
+            LArDiscreteProbabilityHelper::MakeRandomisedSample(t1,randomNumberGenerator),
+            LArDiscreteProbabilityHelper::MakeRandomisedSample(t2,randomNumberGenerator)));
+
+        if ((rRandomised-rNominal) > std::numeric_limits<float>::epsilon())
+            nExtreme++;
+    }
+
+    return static_cast<float>(nExtreme)/static_cast<float>(nPermutations);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename T>
+float LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromStudentTDistribution(const T &t1, 
+    const T &t2, const size_t nIntegrationSteps)
+{
+    const float correlation(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(t1,t2));
+    const float dof(static_cast<float>(LArDiscreteProbabilityHelper::GetSize(t1)) - 2.f);
+    const float tTestStatistic(correlation*sqrt(dof)/(sqrt(1.f - correlation*correlation)));
+    const float tDistCoeff(std::tgamma(0.5f*(dof + 1.f))/std::tgamma(0.5f*dof)/(std::sqrt(dof*M_PI)));
+
+    const float upperLimit(15.f);
+    const float dx((upperLimit - tTestStatistic)/static_cast<float>(nIntegrationSteps));
+    float integral(tDistCoeff*std::pow(1.f + tTestStatistic*tTestStatistic/dof, -0.5f*(dof + 1.f)) + 
+            tDistCoeff*std::pow(1.f + upperLimit*upperLimit/dof, -0.5f*(dof + 1.f)));
+    for (unsigned int iStep = 1; iStep < nIntegrationSteps; ++iStep)
+        integral+=2.f*tDistCoeff*std::pow( 
```

Comment:
Done.  Also changed the final line to be return integral*dx/2.f (rather than calculate then return).

This code isn't performance critical; it isn't actually used anymore since swapping to the permutation test.  It was still quick enough though.